File: crates/dewasm-backend-perl/units/wasi/path_filestat_get.pl
# requires: memory/read_string, memory/init, wasi/resolve_path, wasi/pack_filestat, wasi/errno_fs
use Time::HiRes ();

sub wasi_path_filestat_get {
    my ($self, $dirfd, $flags, $path_ptr, $path_len, $buf_ptr) = @_;
    my $rel = $self->{memory}->read_string($path_ptr, $path_len);
    my $follow = ($flags & 0x1) != 0;  # lookupflags::SYMLINK_FOLLOW
    my ($host, $err) = $self->resolve_path($dirfd, $rel, $follow);
    return $err if defined $err;
    my @st = $follow ? Time::HiRes::stat($host) : Time::HiRes::lstat($host);
    return $self->fs_errno(0 + $!) unless @st;
    $self->{memory}->init($buf_ptr, $self->pack_filestat(\@st), 0, 64);
    return ERRNO_SUCCESS;
}

File: crates/dewasm-backend-go/units/wasi/path_filestat_get.go
// requires: memory/read_string, memory/init, wasi/resolve_path, wasi/pack_filestat, wasi/errno_fs
func (w *WASI) wasi_path_filestat_get(dirfd, flags, pathPtr, pathLen, bufPtr uint32) uint32 {
    rel := string(w.memory.read_string(uint64(pathPtr), uint64(pathLen)))
    symlinkFollow := flags&0x1 != 0 // lookupflags::SYMLINK_FOLLOW
    hostPath, err := w.resolve_path(dirfd, rel, symlinkFollow)
    if err != wasiOk {
        return err
    }
    var fi os.FileInfo
    var e error
    if symlinkFollow {
        fi, e = os.Stat(hostPath)
    } else {
        fi, e = os.Lstat(hostPath)
    }
    if e != nil {
        return w.fs_errno(e)
    }
    w.memory.init(uint64(bufPtr), w.pack_filestat(fi), 0, 64)
    return wasiOk
}
